doc/man/pam_get_authtok.3
.\"-
.\" Copyright (c) 2002 Networks Associates Technologies, Inc.
.\" All rights reserved.
.\"
.\" This software was developed for the FreeBSD Project by ThinkSec AS and
.\" NAI Labs, the Security Research Division of Network Associates, Inc.
.\" under DARPA/SPAWAR contract N66001-01-C-8035 ("CBOSS"), as part of the
.\" DARPA CHATS research program.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $Id$
.\"
.Dd February 4, 2002
.Dt PAM_GET_AUTHTOK 3
.Os
.Sh NAME
.Nm pam_get_authtok
.Nd retrieve authentication token
.Sh LIBRARY
.Lb libpam
.Sh SYNOPSIS
.In security/pam_appl.h
.Ft int
.Fn pam_get_authtok "pam_handle_t *pamh" "const char **authtok" "const char *prompt"
.Sh DESCRIPTION
The
.Nm
function is not yet documented.
.Sh RETURN VALUES
The
.Fn
function returns one of the following values:
.Bl -tag -width PAM_AUTHTOK_DISABLE_AGING
.El
.Sh SEE ALSO
.Xr pam_strerror 3 ,
.Xr pam 3
.Sh STANDARDS
.Rs
.%T "X/Open Single Sign-On Service (XSSO) - Pluggable Authentication Modules"
.%D "June 1997"
.Re
.Sh AUTHORS
The
.Nm
function and this manual page were developed for the FreeBSD Project
by ThinkSec AS and NAI Labs, the Security Research Division of Network
Associates, Inc.  under DARPA/SPAWAR contract N66001-01-C-8035
.Pq Dq CBOSS ,
as part of the DARPA CHATS research program.


HISTORY
OpenPAM Lycopsida						2011-??-??

 - ENHANCE: removed static build autodetection, which didn't work anyway.
   Use an explicit, user-specified preprocessor variable instead.

 - ENHANCE: cleaned up the documentation a bit.

 - ENHANCE: added openpam_subst(3), allowing certain PAM items to be
   embedded in strings such as prompts.  Apply it to the prompts used
   by pam_get_user(3) and pam_get_authtok(3).

 - ENHANCE: add support for the user_prompt, authtok_prompt and
   oldauthtok_prompt module options, which override the prompts passed
   by the module to pam_set_user(3) and pam_get_authtok(3).
============================================================================
OpenPAM Hydrangea						2007-12-21

 - ENHANCE: when compiling with GCC, mark up API functions with GCC
   attributes where appropriate.

 - BUGFIX: fixed numerous warnings uncovered by GCC 4.

 - ENHANCE: building the documentation is now optional.

 - ENHANCE: corrected a number of mistakes and style issues in the
   build system.

 - ENHANCE: API function arguments are now const where appropriate, to
   match corresponding changes in the Solaris PAM and Linux-PAM APIs.

 - ENHANCE: corrected a number of C namespace violations.

 - ENHANCE: the module cache has been removed, allowing long-lived
   applications to pick up module changes.  This also allows multiple
   threads to use PAM simultaneously (as long as they use separate PAM
   contexts), since the module cache was the only part of OpenPAM that
   was not thread-safe.
============================================================================
OpenPAM Figwort							2005-06-16

 - BUGFIX: Correct several small signedness and initialization bugs
   discovered during review by the NetBSD team.

 - BUGFIX: Modify gendoc.pl to sort cross-references in dictionary
   order within each section.

 - ENHANCE: if a policy specifies a relative module path, prepend the
   module directory so we never call dlopen(3) with a relative path.

 - ENHANCE: add a pam.conf(5) manual page.
============================================================================
OpenPAM Feterita						2005-02-01

 - BUGFIX: Correct numerous markup errors, invalid cross-references,
   and other issues in the manual pages, with kind assistance from
   Ruslan Ermilov <ru@freebsd.org>.

 - BUGFIX: Avoid multiple evaluation of macro arguments in ENTERX()
   and RETURNX() macros.

 - BUGFIX: Remove an unnecessary and non-portable pointer cast in
   pam_get_data(3).

 - BUGFIX: Fix identical typos in PAM_ACCT_EXPIRED case in
   pam_strerror(3) and gendoc.pl.

 - ENHANCE: Minor overhaul of the autoconf / build system.

 - ENHANCE: Add openpam_free_envlist(3).
============================================================================
OpenPAM Eelgrass						2004-02-10

 - BUGFIX: Correct array handling bugs in conversation code.

 - BUGFIX: In openpam_ttyconv(3), don't strip trailing linear
   whitespace from the user's response.

 - BUGFIX: Many constness issues addressed.
============================================================================
OpenPAM Dogwood							2003-07-15

 - ENHANCE: Use the GNU autotools.

 - ENHANCE: Constify the msg field in struct pam_message.

 - BUGFIX: Remove left-over debugging output

 - BUGFIX: Avoid side effects in arguments to the FREE() macro

 - ENHANCE: Make openpam_ttyconv(3) use read(2) rather than fgets(3).

 - BUGFIX: Staticize some variables which shouldn't be global.

 - BUGFIX: Correcly anticipate a NULL user in pam_get_user(3).

 - ENHANCE: Various minor documentation improvements.

Thanks to Dmitry V. Levin <ldv@altlinux.org> for considerable
assistance with this release.
============================================================================
OpenPAM Digitalis						2003-06-01

 - ENHANCE: Completely rewrite the configuration parser and add
   support for the "include" control flag.

 - ENHANCE: Improve portability to NetBSD, OpenBSD and Linux.

 - ENHANCE: Lots of additional paranoia.

 - BUGFIX: The sample su(1) application dropped privileges before
   forking instead of after.

 - ENHANCE: Document openpam_log(3).

 - ENHANCE: Other minor documentation fixes.

Thanks to Dmitry V. Levin <ldv@altlinux.org> for considerable
assistance with this release.
============================================================================
OpenPAM Dianthus						2003-05-02

 - BUGFIX: Initialize some potentially uninitialized variables.

 - BUGFIX: Silence some warnings emitted by gcc -std=iso9899:1999.

 - BUGFIX: In pam_getenv(), return a pointer to the stored variable
   instead of a freshly allocated copy.

 - ENHANCE: Detect recursion in openpam_borrow_cred()

 - ENHANCE: Make borrowing one's own credentials a no-op.

 - ENHANCE: Further improve debugging support.

 - ENHANCE: Clean up some variable names.
============================================================================
OpenPAM Daffodil						2003-01-06

 - ENHANCE: Document dependency on <sys/types.h> (for size_t)

 - ENHANCE: Slightly improve error detection in openpam_ttyconv().

 - BUGFIX: Fix several typos in debugging macros.
============================================================================
OpenPAM Cyclamen						2002-12-12

 - ENHANCE: Improve recursion detection in openpam_dispatch().

 - ENHANCE: Add debugging messages at entry and exit points of most
   functions.

 - ENHANCE: Fix some minor style issues.

 - BUGFIX: Add default cases to the switches in openpam_log.c.

 - ENHANCE: Add /usr/local/etc/pam.conf to policy search path.

 - BUGFIX: In openpam_ttyconv(3), print the prompt to stdout rather
   than stderr.
============================================================================
OpenPAM Citronella						2002-06-30

 - ENHANCE: Add the "binding" control flag (from Solaris 9).

 - ENHANCE: Define struct pam_repository and PAM_REPOSITORY (from
   Solaris 9).

 - ENHANCE: Flesh out the pam(3) man page.

 - ENHANCE: Add an openpam(3) page with cross-references to all the

lib/libpam/pam_get_authtok.c
/*-
 * Copyright (c) 2002-2003 Networks Associates Technology, Inc.
 * Copyright (c) 2004-2017 Dag-Erling Smørgrav
 * All rights reserved.
 *
 * This software was developed for the FreeBSD Project by ThinkSec AS and
 * Network Associates Laboratories, the Security Research Division of
 * Network Associates, Inc.  under DARPA/SPAWAR contract N66001-01-C-8035
 * ("CBOSS"), as part of the DARPA CHATS research program.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id$
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <sys/param.h>

#include <stdlib.h>
#include <string.h>

#include <security/pam_appl.h>
#include <security/openpam.h>

#include "openpam_impl.h"
#include "openpam_strlset.h"

static const char authtok_prompt[] = "Password:";
static const char authtok_prompt_remote[] = "Password for %u@%h:";
static const char oldauthtok_prompt[] = "Old Password:";
static const char newauthtok_prompt[] = "New Password:";

/*
 * OpenPAM extension
 *
 * Retrieve authentication token
 */

int
pam_get_authtok(pam_handle_t *pamh,
	int item,
	const char **authtok,
	const char *prompt)
{
	char prompt_buf[1024];
	size_t prompt_size;
	const void *oldauthtok, *prevauthtok, *promptp;
	const char *prompt_option, *default_prompt;
	const void *lhost, *rhost;
	char *resp, *resp2;
	int pitem, r, style, twice;

	ENTER();
	if (pamh == NULL || authtok == NULL)
		RETURNC(PAM_SYSTEM_ERR);
	*authtok = NULL;
	twice = 0;
	switch (item) {
	case PAM_AUTHTOK:
		pitem = PAM_AUTHTOK_PROMPT;
		prompt_option = "authtok_prompt";
		default_prompt = authtok_prompt;
		r = pam_get_item(pamh, PAM_RHOST, &rhost);
		if (r == PAM_SUCCESS && rhost != NULL) {
			r = pam_get_item(pamh, PAM_HOST, &lhost);
			if (r == PAM_SUCCESS && lhost != NULL) {
				if (strcmp(rhost, lhost) != 0)
					default_prompt = authtok_prompt_remote;
			}
		}
		r = pam_get_item(pamh, PAM_OLDAUTHTOK, &oldauthtok);
		if (r == PAM_SUCCESS && oldauthtok != NULL) {
			default_prompt = newauthtok_prompt;
			twice = 1;
		}
		break;
	case PAM_OLDAUTHTOK:
		pitem = PAM_OLDAUTHTOK_PROMPT;
		prompt_option = "oldauthtok_prompt";
		default_prompt = oldauthtok_prompt;
		twice = 0;
		break;
	default:
		RETURNC(PAM_SYMBOL_ERR);
	}
	if (openpam_get_option(pamh, "try_first_pass") ||
	    openpam_get_option(pamh, "use_first_pass")) {
		r = pam_get_item(pamh, item, &prevauthtok);
		if (r == PAM_SUCCESS && prevauthtok != NULL) {
			*authtok = prevauthtok;
			RETURNC(PAM_SUCCESS);
		} else if (openpam_get_option(pamh, "use_first_pass")) {
			RETURNC(r == PAM_SUCCESS ? PAM_AUTH_ERR : r);
		}
	}
	/* pam policy overrides the module's choice */
	if ((promptp = openpam_get_option(pamh, prompt_option)) != NULL)
		prompt = promptp;
	/* no prompt provided, see if there is one tucked away somewhere */
	if (prompt == NULL) {
		r = pam_get_item(pamh, pitem, &promptp);
		if (r == PAM_SUCCESS && promptp != NULL)
			prompt = promptp;
	}
	/* fall back to hardcoded default */
	if (prompt == NULL)
		prompt = default_prompt;
	/* expand */
	prompt_size = sizeof prompt_buf;
	r = openpam_subst(pamh, prompt_buf, &prompt_size, prompt);
	if (r == PAM_SUCCESS && prompt_size <= sizeof prompt_buf)
		prompt = prompt_buf;
	style = openpam_get_option(pamh, "echo_pass") ?
	    PAM_PROMPT_ECHO_ON : PAM_PROMPT_ECHO_OFF;
	r = pam_prompt(pamh, style, &resp, "%s", prompt);
	if (r != PAM_SUCCESS)
		RETURNC(r);
	if (twice) {
		r = pam_prompt(pamh, style, &resp2, "Retype %s", prompt);
		if (r != PAM_SUCCESS) {
			strlset(resp, 0, PAM_MAX_RESP_SIZE);
			FREE(resp);
			RETURNC(r);
		}
		if (strcmp(resp, resp2) != 0) {
			strlset(resp, 0, PAM_MAX_RESP_SIZE);
			FREE(resp);
		}
		strlset(resp2, 0, PAM_MAX_RESP_SIZE);
		FREE(resp2);
	}
	if (resp == NULL)
		RETURNC(PAM_TRY_AGAIN);
	r = pam_set_item(pamh, item, resp);
	strlset(resp, 0, PAM_MAX_RESP_SIZE);
	FREE(resp);
	if (r != PAM_SUCCESS)
		RETURNC(r);
	r = pam_get_item(pamh, item, (const void **)authtok);
	RETURNC(r);
}

/*
 * Error codes:
 *
 *	=pam_get_item
 *	=pam_prompt
 *	=pam_set_item
 *	!PAM_SYMBOL_ERR
 *	PAM_TRY_AGAIN
 */

/**
 * The =pam_get_authtok function either prompts the user for an
 * authentication token or retrieves a cached authentication token,
 * depending on circumstances.
 * Either way, a pointer to the authentication token is stored in the
 * location pointed to by the =authtok argument, and the corresponding PAM
 * item is updated.
 *
 * The =item argument must have one of the following values:
 *
 *	=PAM_AUTHTOK:
 *		Returns the current authentication token, or the new token
 *		when changing authentication tokens.
 *	=PAM_OLDAUTHTOK:
 *		Returns the previous authentication token when changing
 *		authentication tokens.
 *
 * The =prompt argument specifies a prompt to use if no token is cached.
 * If it is =NULL, the =PAM_AUTHTOK_PROMPT or =PAM_OLDAUTHTOK_PROMPT item,
 * as appropriate, will be used.
 * If that item is also =NULL, a hardcoded default prompt will be used.
 * Additionally, when =pam_get_authtok is called from a service module,
 * the prompt may be affected by module options as described below.
 * The prompt is then expanded using =openpam_subst before it is passed to
 * the conversation function.
 *
 * If =item is set to =PAM_AUTHTOK and there is a non-null =PAM_OLDAUTHTOK
 * item, =pam_get_authtok will ask the user to confirm the new token by
 * retyping it.
 * If there is a mismatch, =pam_get_authtok will return =PAM_TRY_AGAIN.
 *
 * MODULE OPTIONS
 *
 * When called by a service module, =pam_get_authtok will recognize the
 * following module options:
 *
 *	;authtok_prompt:
 *		Prompt to use when =item is set to =PAM_AUTHTOK.
 *		This option overrides both the =prompt argument and the
 *		=PAM_AUTHTOK_PROMPT item.
 *	;echo_pass:
 *		If the application's conversation function allows it, this
 *		lets the user see what they are typing.
 *		This should only be used for non-reusable authentication
 *		tokens.
 *	;oldauthtok_prompt:
 *		Prompt to use when =item is set to =PAM_OLDAUTHTOK.
 *		This option overrides both the =prompt argument and the
 *		=PAM_OLDAUTHTOK_PROMPT item.
 *	;try_first_pass:
 *		If the requested item is non-null, return it without
 *		prompting the user.
 *		Typically, the service module will verify the token, and
 *		if it does not match, clear the item before calling
 *		=pam_get_authtok a second time.
 *	;use_first_pass:
 *		Do not prompt the user at all; just return the cached
 *		value, or =PAM_AUTH_ERR if there is none.
 *
 * >pam_conv
 * >pam_get_item
 * >pam_get_user
 * >openpam_get_option
 * >openpam_subst
 */
